baseplate/diagnostics/tracing.py

```diff
@@ -0,0 +1,226 @@
+"""Components for processing Baseplate spans for service request tracing."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import calendar
+import json
+import logging
+import socket
+import time
+from datetime import datetime
+
+import gevent
+import requests
+
+from ..core import BaseplateObserver, SpanObserver, TraceInfo
+
+
+logger = logging.getLogger(__name__)
+
+
+# Span annotation types
+ANNOTATIONS = {
+    'CLIENT_SEND': 'cs',
+    'CLIENT_RECEIVE': 'cr',
+    'SERVER_SEND': 'ss',
+    'SERVER_RECEIVE': 'sr',
+}
+
+
+def current_epoch_microseconds():
+    """Return current UTC time since epoch in microseconds."""
+    epoch_ts = datetime.utcfromtimestamp(0)
+    return int((datetime.utcnow() - epoch_ts).
+               total_seconds() * 1000 * 1000)
```

Comment:
I have a collection of functions in the gateway that do things like this, I wonder if it would be better to collect them into a util in baseplate?  They look like this:

```python
EPOCH = datetime(1970, 1, 1, tzinfo=pytz.UTC)


def utc_now():
    return datetime.now(pytz.UTC)


def epoch_timestamp(dt):
    """Returns the number of seconds from the epoch to date."""
    return (dt - EPOCH).total_seconds()


def to_epoch_milliseconds(dt):
    """Returns the number of milliseconds from the epoch to date."""
    return int(math.floor(epoch_timestamp(dt) * 1000.))


def from_epoch_milliseconds(ms):
    """Convert milliseconds from the epoch to UTC datetime.

    :param int ms: milliseconds since the epoch
    :rtype: :py:class:`datetime.datetime`
    """
    seconds = int(ms / 1000.)
    microseconds = (ms - 1000 * seconds) * 1000.
    return EPOCH + timedelta(seconds=seconds, microseconds=microseconds)
```